### PR TITLE
Good edge cases

### PR DESCRIPTION
**Pull request by DanielKrawisz**
I wrote a bunch of tests. They all build on one another, so there are several commits. I tested a bunch of weird edge cases. All of these can be sent to btdc too, so I'll do that eventually.

### CODE REVIEW

File: addrmgr/addrmanager.go [hunk 1]
```diff
@@ -741,7 +741,9 @@ func (a *AddrManager) GetAddress(class string, newBias int) *KnownAddress {
 		(100.0 - float64(newBias))
 	newCorrelation := math.Sqrt(float64(a.nNew)) * float64(newBias)
 
-	if ((newCorrelation + triedCorrelation) * a.rand.Float64()) <
+	//If there are tried vectors but no new vectors, automatically select from tried
```
Comment (ishbir) [inline addrmgr/addrmanager.go]: Very minor but it's usually nice to leave a space between // and the first letter.

File: addrmgr/internal_test.go [hunk 2]
```diff
@@ -16,5 +16,133 @@ func KnownAddressChance(ka *KnownAddress) float64 {
 
 func NewKnownAddress(na *wire.NetAddress, attempts int,
```
Comment (ishbir) [inline addrmgr/internal_test.go]: The convention used in btcd and bmd is to prefix all functions in internal_test.go with Tst. Please make sure to follow that. I forgot to point this out in your previous commit.

File: addrmgr/addrmanager_test.go [hunk 3]
```diff
@@ -353,6 +405,44 @@ func TestNeedMoreAddresses(t *testing.T) {
...
+	n.AddAddressesSkipChecks(addrs, srcAddr)
+	expected := n.NumAddresses() * addrmgr.GetAddrPercent() / 100
+	cache = n.AddressCache()
+	if cache == nil {
+		t.Error("Address cache should not be nil.")
+	} else if len(cache) != expected {
+		t.Errorf("Address cache should contain %d addresses but it has %d.", expected, len(cache))
+	}
+
+	//More than getAddrMax addresses have been added.
```
Comment (ishbir) [inline addrmgr/addrmanager_test.go]: Same here.
